File: cli-os/internal/gateway/upstream.go
// Low-level upstream call helpers shared by the ingress streaming path and the internal turn
// primitive. Extracted so a bridge hop reuses the exact same idempotency-aware, abort-aware call
// path as a top-level request — never by re-entering HTTP. Ported from upstream.js. Go's
// context.WithTimeout(clientCtx, ...) gives both the request timeout and client-disconnect abort
// (the clientCtx is cancelled when the client closes the connection), covering the full body read.
package gateway

import (
	"bytes"
	"context"
	"encoding/json"
	"fmt"
	"io"
	"net/http"
	"strings"
	"time"

	"github.com/jackofall1232/l00prite/cli-os/internal/config"
	"github.com/jackofall1232/l00prite/cli-os/internal/gateway/adapters"
)

// httpClient uses the default transport, which honors HTTPS_PROXY and the system CA bundle. No
// client-level timeout — the per-request context deadline governs (and covers streaming body reads).
var httpClient = &http.Client{}

func isRetryable(status int) bool { return status == 429 || status == 408 || status >= 500 }

// providerFetch issues the POST under a context that both times out and cancels on client
// disconnect. On success the caller MUST call the returned cancel() after the body is fully read; on
// error cancel is a safe no-op (already cancelled).
func providerFetch(ctx context.Context, url string, headers map[string]string, body any, timeoutMs int) (*http.Response, context.CancelFunc, error) {
	cctx, cancel := context.WithTimeout(ctx, time.Duration(timeoutMs)*time.Millisecond)
	b, err := json.Marshal(body)
	if err != nil {
		cancel()
		return nil, func() {}, err
	}
	req, err := http.NewRequestWithContext(cctx, http.MethodPost, url, bytes.NewReader(b))
	if err != nil {
		cancel()
		return nil, func() {}, err
	}
	for k, v := range headers {
		req.Header.Set(k, v)
	}
	resp, err := httpClient.Do(req)
	if err != nil {
		cancel()
		return nil, func() {}, err
	}
	return resp, cancel, nil
}

func readLimited(r io.Reader, n int64) string {
	b, _ := io.ReadAll(io.LimitReader(r, n))
	return string(b)
}

func truncStr(s string, n int) string {
	if len(s) <= n {
		return s
	}
	r := []rune(s)
	if len(r) <= n {
		return s
	}
	return string(r[:n]) // cap at n runes, never splitting a multibyte character
}

func backoff(baseMs, attempt, maxMs int) time.Duration {
	ms := baseMs << attempt
	if ms > maxMs {
		ms = maxMs
	}
	return time.Duration(ms) * time.Millisecond
}

// callNonStream is the non-streaming provider call with idempotency-aware retry (no client-visible
// bytes have flushed, so every attempt within the cap is safe).
func callNonStream(a adapters.Adapter, provider ProviderRow, model string, req map[string]any, apiKey string, cfg config.Config, clientCtx context.Context) (adapters.FullResult, error) {
	if a.Direct() {
		return a.(adapters.DirectAdapter).DirectFull(req, model), nil
	}
	na := a.(adapters.NetworkAdapter)
	body := na.BuildRequest(model, req, false)
	url := na.URL(provider.BaseURL)
	headers := na.Headers(apiKey)

	var lastErr error
	for attempt := 0; attempt < cfg.Retry.MaxAttempts; attempt++ {
		resp, cancel, err := providerFetch(clientCtx, url, headers, body, cfg.RequestTimeoutMs)
		if err != nil {
			if clientCtx.Err() != nil {
				return adapters.FullResult{}, HTTPError(499, "client closed request", "client_closed")
			}
			lastErr = err
		} else if resp.StatusCode >= 200 && resp.StatusCode < 300 {
			var j map[string]any
			derr := json.NewDecoder(resp.Body).Decode(&j)
			resp.Body.Close()
			cancel()
			if derr == nil {
				return na.ParseFull(j, model), nil
			}
			lastErr = HTTPError(502, "upstream "+provider.Name+" returned invalid JSON", "upstream_error")
		} else {
			errText := readLimited(resp.Body, 4096)
			resp.Body.Close()
			cancel()
			if !isRetryable(resp.StatusCode) {
				return adapters.FullResult{}, HTTPError(resp.StatusCode,
					fmt.Sprintf("upstream %s %d: %s", provider.Name, resp.StatusCode, truncStr(errText, 300)), "upstream_error")
			}
			lastErr = HTTPError(502, fmt.Sprintf("upstream %s %d", provider.Name, resp.StatusCode), "upstream_error")
		}
		if clientCtx.Err() != nil {
			return adapters.FullResult{}, HTTPError(499, "client closed request", "client_closed")
		}
		time.Sleep(backoff(cfg.Retry.BaseMs, attempt, cfg.Retry.MaxMs))
	}
	if lastErr != nil {
		return adapters.FullResult{}, lastErr
	}
	return adapters.FullResult{}, HTTPError(502, "upstream failed", "upstream_error")
}

// parseSSE parses complete SSE frames from a buffer already CRLF-normalized to \n. It yields every
// frame except the trailing (possibly incomplete) one, matching the JS generator's semantics.
func parseSSE(buffer string) []adapters.SSEEvent {
	parts := strings.Split(buffer, "\n\n")
	var out []adapters.SSEEvent
	for i := 0; i < len(parts)-1; i++ {
		var event string
		var dataLines []string
		for _, line := range strings.Split(parts[i], "\n") {
			switch {
			case strings.HasPrefix(line, "event:"):
				event = strings.TrimSpace(line[len("event:"):])
			case strings.HasPrefix(line, "data:"):
				dataLines = append(dataLines, strings.TrimSpace(line[len("data:"):]))
			}
		}
		if len(dataLines) > 0 {
			out = append(out, adapters.SSEEvent{Event: event, Data: strings.Join(dataLines, "\n")})
		}
	}
	return out
}

File: cli-os/src/gateway/upstream.js
// Low-level upstream call helpers shared by the ingress streaming path and the internal turn
// primitive (turn.js). Extracted so a bridge hop reuses the exact same idempotency-aware,
// abort-aware call path as a top-level request — never by re-entering HTTP (which would re-read
// bridge headers and re-inject the bridge tool: the recursion hole Fable flagged).
import * as router from './router.js';

export const sleep = (ms) => new Promise((r) => setTimeout(r, ms));
export const isRetryable = (status) => status === 429 || status === 408 || status >= 500;

// Returns { promise, cleanup }. cleanup() clears the timeout — call it only AFTER the response
// body/stream has been fully consumed, so a stalled body still aborts on timeout. Aborts on
// client disconnect too.
export function providerFetch(url, headers, body, { timeoutMs, clientSignal }) {
  const ac = new AbortController();
  const timer = setTimeout(() => ac.abort(new Error('provider_timeout')), timeoutMs);
  const onClientAbort = () => ac.abort(new Error('client_disconnected'));
  if (clientSignal) {
    if (clientSignal.aborted) ac.abort();
    else clientSignal.addEventListener('abort', onClientAbort, { once: true });
  }
  const cleanup = () => { clearTimeout(timer); clientSignal?.removeEventListener?.('abort', onClientAbort); };
  const promise = fetch(url, { method: 'POST', headers, body: JSON.stringify(body), signal: ac.signal });
  return { promise, cleanup };
}

// Non-streaming provider call with idempotency-aware retry (a non-streaming call has flushed no
// client-visible bytes, so every attempt is safe to retry within the cap).
export async function callNonStream(adapter, { provider, model, openaiReq, apiKey, cfg, clientSignal }) {
  if (adapter.direct) return adapter.directFull(openaiReq, model); // mock: no network
  const body = adapter.buildRequest({ model, openaiReq, stream: false });
  const url = adapter.url(provider.base_url);
  const headers = adapter.headers(apiKey);
  let lastErr;
  for (let attempt = 0; attempt < cfg.retry.maxAttempts; attempt++) {
    const { promise, cleanup } = providerFetch(url, headers, body, { timeoutMs: cfg.requestTimeoutMs, clientSignal });
    try {
      const r = await promise;
      if (r.ok) return adapter.parseFull(await r.json(), model);
      const errText = await r.text().catch(() => '');
      if (!isRetryable(r.status)) throw router.httpError(r.status, `upstream ${provider.name} ${r.status}: ${errText.slice(0, 300)}`, 'upstream_error');
      lastErr = router.httpError(502, `upstream ${provider.name} ${r.status}`, 'upstream_error');
    } catch (e) {
      if (clientSignal?.aborted) throw router.httpError(499, 'client closed request', 'client_closed');
      if (e.status && !isRetryable(e.status)) throw e;
      lastErr = e;
    } finally { cleanup(); }
    await sleep(Math.min(cfg.retry.baseMs * 2 ** attempt, cfg.retry.maxMs));
  }
  throw lastErr || router.httpError(502, 'upstream failed', 'upstream_error');
}

// SSE frame parser (buffer is already CRLF-normalized to \n).
export function* parseSSE(buffer) {
  const parts = buffer.split('\n\n');
  for (let i = 0; i < parts.length - 1; i++) {
    let event = null; const dataLines = [];
    for (const line of parts[i].split('\n')) {
      if (line.startsWith('event:')) event = line.slice(6).trim();
      else if (line.startsWith('data:')) dataLines.push(line.slice(5).trim());
    }
    if (dataLines.length) yield { event, data: dataLines.join('\n') };
  }
}
